File: CutAndShake.glyphsFilter/Contents/Resources/CutAndShake.py
#!/usr/bin/env python
# encoding: utf-8

import objc
from Foundation import *
from AppKit import *
import sys, os, re
import random
import math

class GlyphsFilterCutAndShake ( GSFilterPlugin ):
	goodMeasure = 5.0
	
	_numberOfCutsField = objc.IBOutlet()
	_maxMoveField = objc.IBOutlet()
	_maxRotateField = objc.IBOutlet()
	
	def init( self ):
		"""
		Do all initializing here.
		"""
		NSBundle.loadNibNamed_owner_( "CutAndShakeDialog", self )
		random.seed()
		return self
	
	def interfaceVersion( self ):
		"""
		Distinguishes the API version the plugin was built for. 
		Return 1.
		"""
		return 1
	
	def title( self ):
		"""
		This is the human-readable name as it appears in the menu.
		"""
		try:
			return "Cut and Shake"
		except Exception as e:
			self.logToConsole( "title: %s" % str(e) )
	
	def actionName( self ):
		"""
		This is the title of the button in the settings dialog.
		"""
		try:
			return "Apply"
		except Exception as e:
			self.logToConsole( "actionName: %s" % str(e) )
	
	def keyEquivalent( self ):
		""" 
		The key together with Cmd+Shift will be the shortcut for the filter.
		Return None if you do not want to set a shortcut.
		Users can set their own shortcuts in System Prefs.
		"""
		try:
			return None
		except Exception as e:
			self.logToConsole( "keyEquivalent: %s" % str(e) )
	
	def setup(self):
		try:
			super( GlyphsFilterCutAndShake, self ).setup()
			FontMaster = self.valueForKey_( "fontMaster" )
		
			self.numberOfCuts = self.setDefaultIntegerValue( "numberOfCuts", 20, FontMaster )
			self.maxMove   = self.setDefaultFloatValue( "maxMove",  20.0, FontMaster )
			self.maxRotate = self.setDefaultFloatValue( "maxRotate", 8.0, FontMaster )
		
			self._numberOfCutsField.setIntegerValue_( self.numberOfCuts )
			self._maxMoveField.setFloatValue_( self.maxMove )
			self._maxRotateField.setFloatValue_( self.maxRotate )
		
			self.process_( None )
			return None # or if something goes wrong, a NSError object with details
		except Exception as e:
			self.logToConsole( "setup: %s" % str(e) )
	
	def setDefaultFloatValue( self, userDataKey, defaultValue, FontMaster ):
		"""Returns either the stored or default value for the given userDataKey."""
		if userDataKey in FontMaster.userData:
			return FontMaster.userData[userDataKey].floatValue()
		else:
			return defaultValue
			
	def setDefaultIntegerValue( self, userDataKey, defaultValue, FontMaster ):
		"""Returns either the stored or default value for the given userDataKey."""
		if userDataKey in FontMaster.userData:
			return FontMaster.userData[userDataKey].integerValue()
		else:
			return defaultValue
	
	@objc.IBAction
	def setNumberOfCutsValue_( self ,sender ):
		numberOfCuts = sender.floatValue()
		if numberOfCuts != self.numberOfCuts:
			self.numberOfCuts = numberOfCuts
			self.process_( None )

	@objc.IBAction
	def setMaxMoveValue_( self ,sender ):
		maxMove = sender.floatValue()
		if maxMove != self.maxMove:
			self.maxMove = maxMove
			self.process_( None )
			
	@objc.IBAction
	def setMaxRotateValue_( self ,sender ):
		maxRotate = sender.floatValue()
		if maxRotate != self.maxRotate:
			self.maxRotate = maxRotate
			self.process_( None )
	
	def processLayerWithValues( self, Layer, numberOfCuts, maxMove, maxRotate ):
		"""
		This is where your code for processing each layer goes.
		This method is the one eventually called by either the Custom Parameter or Dialog UI.
		"""
		try:
			self.randomCutLayer( Layer, numberOfCuts )
			self.randomMovePaths( Layer, abs(maxMove) )
			self.randomRotatePaths( Layer, abs(maxRotate) )
		except Exception as e:
			self.logToConsole( "processLayerWithValues: %s" % str(e) )
	
	def processFont_withArguments_( self, Font, Arguments ):
		"""
		Invoked when called as Custom Parameter in an instance at export.
		The Arguments come from the custom parameter in the instance settings. 
		The first item in Arguments is the class-name. After that, it depends on the filter.
		"""
		try:
			# Set default values for potential arguments (values), just in case:
			numberOfCuts = 10
			maxMove = 30.0
			maxRotate = 20.0
			
			# Override defaults with actual values from custom parameter:
			if len(Arguments) > 1:
				numberOfCuts = Arguments[1].integerValue()
				maxMove = Arguments[2].floatValue()
				maxRotate = Arguments[3].floatValue()
		
			# With these values, call our code on every glyph:
			FontMasterId = Font.fontMasterAtIndex_(0).id
			for Glyph in Font.glyphs:
				Layer = Glyph.layerForKey_( FontMasterId )
				self.processLayerWithValues( Layer, numberOfCuts, maxMove, maxRotate )
		except Exception as e:
			self.logToConsole( "processFont_withArguments_: %s" % str(e) )
	
	def process_( self, sender ):
		"""
		This method gets called when the filter is run with the dialog through the Filter menu.
		"""
		try:
			# Create Preview in Edit View, save and show original in ShadowLayer:
			ShadowLayers = self.valueForKey_( "shadowLayers" )
			Layers = self.valueForKey_( "layers" )
			checkSelection = True
			for k in range(len( ShadowLayers )):
				ShadowLayer = ShadowLayers[k]
				Layer = Layers[k]
				Layer.setPaths_( NSMutableArray.alloc().initWithArray_copyItems_( ShadowLayer.pyobjc_instanceMethods.paths(), True ) )
				Layer.setSelection_( NSMutableArray.array() )
				if len(ShadowLayer.selection()) > 0 and checkSelection:
					for i in range(len( ShadowLayer.paths )):
						currShadowPath = ShadowLayer.paths[i]
						currLayerPath = Layer.paths[i]
						for j in range(len(currShadowPath.nodes)):
							currShadowNode = currShadowPath.nodes[j]
							if ShadowLayer.selection().containsObject_( currShadowNode ):
								Layer.addSelection_( currLayerPath.nodes[j] )
				self.processLayerWithValues( Layer, self.numberOfCuts, self.maxMove, self.maxRotate )
			Layer.clearSelection()
		
			# Safe the values in the FontMaster. But could be saved in UserDefaults, too.
			FontMaster = self.valueForKey_( "fontMaster" )
			FontMaster.userData[ "numberOfCuts" ] = NSNumber.numberWithInteger_( self.numberOfCuts )
			FontMaster.userData[ "maxMove" ] = NSNumber.numberWithDouble_( self.maxMove )
			FontMaster.userData[ "maxRotate" ] = NSNumber.numberWithDouble_( self.maxRotate )
			
			# call the superclass to trigger the immediate redraw:
			super( GlyphsFilterCutAndShake, self ).process_( sender )
		except Exception as e:
			self.logToConsole( "process_: %s" % str(e) )
	
	def logToConsole( self, message ):
		"""
		The variable 'message' will be passed to Console.app.
		Use self.logToConsole( "bla bla" ) for debugging.
		"""
		myLog = "Filter %s:\n%s" % ( self.title(), message )
		NSLog( myLog )
		
	def rotationTransform( self, angle=180.0, x_orig=0.0, y_orig=0.0 ):
		"""Returns a TransformStruct for rotating."""
		try:
			RotationTransform = NSAffineTransform.transform()
			RotationTransform.translateXBy_yBy_( x_orig, y_orig )
			RotationTransform.rotateByDegrees_( angle )
			RotationTransform.translateXBy_yBy_( -x_orig, -y_orig )
			return RotationTransform
		except Exception as e:
			self.logToConsole( "rotationTransform: %s" % str(e) )

	def translateTransform( self, x_move=0.0, y_move=0.0 ):
		"""Returns a TransformStruct for translating."""
		try:
			TranslateTransform = NSAffineTransform.transform()
			TranslateTransform.translateXBy_yBy_( x_move, y_move )
			return TranslateTransform
		except Exception as e:
			self.logToConsole( "translateTransform: %s" % str(e) )

	def somewhereBetween( self, minimum, maximum ):
		try:
			precisionFactor = 100.0
			return random.randint( int( minimum * precisionFactor ), int( maximum * precisionFactor ) ) / precisionFactor
		except Exception as e:
			self.logToConsole( "somewhereBetween: %s" % str(e) )

	def rotate( self, x, y, angle=180.0, x_orig=0.0, y_orig=0.0):
		"""Rotates x/y around x_orig/y_orig by angle and returns result as [x,y]."""
		# TO DO: update this to use self.rotationTransform()
		try:
			new_angle = ( angle / 180.0 ) * math.pi
			new_x = ( x - x_orig ) * math.cos( new_angle ) - ( y - y_orig ) * math.sin( new_angle ) + x_orig
			new_y = ( x - x_orig ) * math.sin( new_angle ) + ( y - y_orig ) * math.cos( new_angle ) + y_orig
			return [ new_x, new_y ]
		except Exception as e:
			self.logToConsole( "rotate: %s" % str(e) )

	def rotatePath( self, thisPath, angle=180.0, x_orig=0.0, y_orig=0.0 ):
		"""Rotates a path around x_orig/y_orig by angle."""
		# TO DO: update this to use self.rotationTransform()
		try:
			for thisNode in thisPath.nodes:
				[ thisNode.x, thisNode.y ] = self.rotate( thisNode.x, thisNode.y, angle=(angle/1.0), x_orig=x_orig, y_orig=y_orig )
		except Exception as e:
			self.logToConsole( "rotatePath: %s" % str(e) )

	def translatePath( self, thisPath, translateByX=0.0, translateByY=0.0 ):
		"""Translates the path by translateByX/translateByY."""
		try:
			for thisNode in thisPath.nodes:
				thisNode.x += translateByX
				thisNode.y += translateByY
		except Exception as e:
			self.logToConsole( "translatePath: %s" % str(e) )

	def randomCutLayer( self, thisLayer, numberOfCuts ):
		try:
			lowestY = thisLayer.bounds.origin.y - self.goodMeasure
			highestY = thisLayer.bounds.origin.y + thisLayer.bounds.size.height + self.goodMeasure
			leftmostX = thisLayer.bounds.origin.x - self.goodMeasure
			rightmostX = thisLayer.bounds.origin.x + thisLayer.bounds.size.width + self.goodMeasure
			for i in range( numberOfCuts ):
				# make either horizontal or vertical cut:
				if random.randint( 0, 1 ) == 0:
					pointL = NSPoint( leftmostX,  self.somewhereBetween( lowestY, highestY ) )
					pointR = NSPoint( rightmostX, self.somewhereBetween( lowestY, highestY ) )
					thisLayer.cutBetweenPoints( pointL, pointR )
				else:
					pointB = NSPoint( self.somewhereBetween( leftmostX, rightmostX ), lowestY  )
					pointT = NSPoint( self.somewhereBetween( leftmostX, rightmostX ), highestY )
					thisLayer.cutBetweenPoints( pointB, pointT )
			return True
		except Exception as e:
			self.logToConsole( "randomCutLayer: %s" % str(e) )

	def randomMovePaths( self, thisLayer, maximumMove ):
		try:
			for thisPath in thisLayer.paths:
				xMove = self.somewhereBetween( -maximumMove/(2**0.5), maximumMove/(2**0.5) )
				yMove = self.somewhereBetween( -maximumMove/(2**0.5), maximumMove/(2**0.5) )
				self.translatePath( thisPath, translateByX=xMove, translateByY=yMove )
				#thisPath.transform = translateTransform( x_move=xMove, y_move=yMove )
			return True
		except Exception as e:
			self.logToConsole( "randomMovePaths: %s" % str(e) )
			return False

	def randomRotatePaths( self, thisLayer, maximumRotate ):
		try:
			for thisPath in thisLayer.paths:
				centerX = thisPath.bounds.origin.x + thisPath.bounds.size.width / 2.0
				centerY = thisPath.bounds.origin.y + thisPath.bounds.size.height / 2.0
				pathRotate = self.somewhereBetween( -maximumRotate, maximumRotate )
				self.rotatePath( thisPath, angle=pathRotate, x_orig=centerX, y_orig=centerY )
				# thisPath.transform = rotationTransform( angle=pathRotate, x_orig=centerX, y_orig=centerY )
			return True
		except Exception as e:
			self.logToConsole( "randomRotatePaths: %s" % str(e) )
			return False
	

File: Cocoa/CutAndShake/CutAndShake.m
//
//  CutAndShake.m
//  CutAndShake
//
//  Rainer Erich Scheichelbauer | mekkablue
//  Native Cocoa/ObjC version of the CutAndShake Glyphs filter plugin.
//
//  Cuts glyphs with random angled lines, then randomly moves and rotates
//  the resulting path fragments.
//

#import "CutAndShake.h"
#import <GlyphsCore/GlyphsCore.h>
#import <GlyphsApp/GSCallbackHandler.h>

// +cutPathsInLayer:forPoint:endPoint: is what GSLayer.cutBetweenPoints() calls
// in the Python wrapper. The class holding it is looked up at runtime, so the
// selector is declared here to give the compiler a signature to work with.
@interface NSObject (GlyphsCutPathsInLayer)
+ (void)cutPathsInLayer:(GSLayer *)layer forPoint:(NSPoint)point endPoint:(NSPoint)endPoint;
@end

@interface CutAndShake ()
+ (void)registerUserDefaults;
- (void)applyFilterToLayer:(GSLayer *)layer
              numberOfCuts:(NSInteger)numberOfCuts
                   maxMove:(CGFloat)maxMove
                 maxRotate:(CGFloat)maxRotate;
@end

// NSUserDefaults keys
static NSString *const kNumberOfCuts = @"com.mekkablue.CutAndShake.numberOfCuts";
static NSString *const kMaxMove      = @"com.mekkablue.CutAndShake.maxMove";
static NSString *const kMaxRotate    = @"com.mekkablue.CutAndShake.maxRotate";

// Extra margin around the glyph bounds when generating cut lines
static const CGFloat kGoodMeasure = 5.0;

/**
 The tool class that performs the cutting.
 Glyphs 4 calls it GlyphsToolKnife, Glyphs 3 called it GlyphsToolOther.
 */
static Class CutPathsToolClass(void) {
	static Class toolClass = Nil;
	static dispatch_once_t onceToken;
	dispatch_once(&onceToken, ^{
		toolClass = NSClassFromString(@"GlyphsToolKnife");
		if (!toolClass) {
			toolClass = NSClassFromString(@"GlyphsToolOther");
		}
	});
	return toolClass;
}

@implementation CutAndShake {
	// Glyphs 4 no longer provides a _view ivar in GSFilterPlugin, so the
	// plugin brings its own. Without it the bundle fails to load in Glyphs 4
	// with a missing _OBJC_IVAR_$_GSFilterPlugin._view symbol.
	NSView *_view;
}

- (instancetype)init {
	self = [super init];
	if (self) {
		[[self class] registerUserDefaults];
	}
	return self;
}

+ (void)registerUserDefaults {
	[[NSUserDefaults standardUserDefaults] registerDefaults:@{
		kNumberOfCuts: @5,
		kMaxMove:      @50,
		kMaxRotate:    @20,
	}];
}

#pragma mark - GSFilterPlugin required methods

- (NSUInteger)interfaceVersion {
	// Distinguishes the API version the plugin was built for.
	return 1;
}

- (NSString *)title {
	// Return the localised menu name.
	// Glyphs picks the best match for the current UI language.
	NSDictionary *names = @{
		@"en": @"Cut and Shake",
		@"de": @"Schneiden und schütteln",
		@"fr": @"Couper et secouer",
		@"es": @"Cortar y agitar",
		@"zh": @"🤺碎片化",
	};
	NSString *lang = [[[NSBundle mainBundle] preferredLocalizations] firstObject] ?: @"en";
	return names[lang] ?: names[@"en"];
}

- (NSString *)actionName {
	// The label of the Apply button in the filter dialog.
	NSDictionary *labels = @{
		@"en": @"Apply",
		@"de": @"Anwenden",
		@"fr": @"Appliquer",
		@"es": @"Aplicar",
		@"zh": @"应用",
	};
	NSString *lang = [[[NSBundle mainBundle] preferredLocalizations] firstObject] ?: @"en";
	return labels[lang] ?: labels[@"en"];
}

- (NSString *)keyEquivalent {
	// Return nil — no fixed keyboard shortcut (users set their own in System Settings).
	return nil;
}

#pragma mark - Dialog / View

- (NSView *)view {
	if (!_view) {
		[[NSBundle bundleForClass:[self class]]
			loadNibNamed:@"IBdialog"
			owner:self
			topLevelObjects:nil];
	}
	return _view;
}

- (NSError *)setup {
	// Called just before the dialog is shown.  Restore saved values and
	// push them into the text fields.
	[super setup];
	[[self class] registerUserDefaults];

	NSUserDefaults *ud = [NSUserDefaults standardUserDefaults];
	_numberOfCutsField.integerValue = [ud integerForKey:kNumberOfCuts];
	_maxMoveField.doubleValue       = [ud doubleForKey:kMaxMove];
	_maxRotateField.doubleValue     = [ud doubleForKey:kMaxRotate];

	// Show an initial preview right away.
	[self process:nil];
	return nil;
}

#pragma mark - IBActions

- (IBAction)setNumberOfCuts:(id)sender {
	[[NSUserDefaults standardUserDefaults]
		setInteger:[(NSTextField *)sender integerValue]
		forKey:kNumberOfCuts];
	[self process:nil];
}

- (IBAction)setMaxMove:(id)sender {
	[[NSUserDefaults standardUserDefaults]
		setDouble:[(NSTextField *)sender doubleValue]
		forKey:kMaxMove];
	[self process:nil];
}

- (IBAction)setMaxRotate:(id)sender {
	[[NSUserDefaults standardUserDefaults]
		setDouble:[(NSTextField *)sender doubleValue]
		forKey:kMaxRotate];
	[self process:nil];
}

#pragma mark - process: (interactive / live preview)

/**
 Called each time the user changes a parameter in the dialog.
 Restores each working layer from the corresponding shadow layer
 (the frozen copy Glyphs made before the filter was first applied),
 then runs the filter, then hands control back to Glyphs via
 [super process:nil].
 */
- (void)process:(id)sender {
	NSUserDefaults *ud = [NSUserDefaults standardUserDefaults];
	NSInteger numberOfCuts = [ud integerForKey:kNumberOfCuts];
	CGFloat   maxMove      = [ud doubleForKey:kMaxMove];
	CGFloat   maxRotate    = [ud doubleForKey:kMaxRotate];

	for (NSUInteger k = 0; k < _shadowLayers.count; k++) {
		GSLayer *shadowLayer = _shadowLayers[k];
		GSLayer *layer       = _layers[k];

		// Restore to the pre-filter state from the shadow copy.
		layer.shapes    = [[NSMutableArray alloc] initWithArray:shadowLayer.shapes copyItems:YES];
		layer.selection = [NSMutableOrderedSet new];

		// Restore selection on individual nodes when the user works
		// in the Edit view with a sub-selection.
		if (shadowLayer.selection.count > 0 && _checkSelection) {
			for (NSUInteger i = 0; i < shadowLayer.shapes.count; i++) {
				GSPath *shadowPath = (GSPath *)[shadowLayer objectInShapesAtIndex:i];
				if (![shadowPath isKindOfClass:[GSPath class]]) continue;
				GSPath *layerPath = (GSPath *)[layer objectInShapesAtIndex:i];
				for (NSUInteger j = 0; j < shadowPath.nodes.count; j++) {
					GSNode *shadowNode = [shadowPath nodeAtIndex:j];
					if ([shadowLayer.selection containsObject:shadowNode]) {
						[layer addSelection:[layerPath nodeAtIndex:j]];
					}
				}
			}
		}

		[self applyFilterToLayer:layer
		            numberOfCuts:numberOfCuts
		                 maxMove:maxMove
		               maxRotate:maxRotate];
		[layer clearSelection];
	}
	[super process:nil];
}

#pragma mark - Custom parameter string

- (NSString *)customParameterString {
	NSUserDefaults *ud = [NSUserDefaults standardUserDefaults];
	return [NSString stringWithFormat:@"%@; cuts:%ld; move:%.1f; rotate:%.1f",
		NSStringFromClass([self class]),
		(long)[ud integerForKey:kNumberOfCuts],
		[ud doubleForKey:kMaxMove],
		[ud doubleForKey:kMaxRotate]];
}

#pragma mark - Export / batch processing

- (void)processFont:(GSFont *)font withArguments:(NSArray *)arguments {
	// Called when the filter is invoked as a Custom Parameter at export.
	// arguments[0] is the class name; the remaining items are key:value pairs.
	[[self class] registerUserDefaults];

	NSUserDefaults *ud = [NSUserDefaults standardUserDefaults];
	NSInteger numberOfCuts = [ud integerForKey:kNumberOfCuts];
	CGFloat   maxMove      = [ud doubleForKey:kMaxMove];
	CGFloat   maxRotate    = [ud doubleForKey:kMaxRotate];

	// Parse key:value arguments supplied via the custom parameter.
	NSCharacterSet *whitespace = [NSCharacterSet whitespaceCharacterSet];
	for (NSUInteger i = 1; i < arguments.count; i++) {
		NSString *arg = [arguments[i] stringByTrimmingCharactersInSet:whitespace];
		if ([arg hasPrefix:@"include:"] || [arg hasPrefix:@"exclude:"]) continue;
		NSArray  *kv  = [arg componentsSeparatedByString:@":"];
		if (kv.count != 2) continue;
		NSString *key   = [kv[0] stringByTrimmingCharactersInSet:whitespace];
		NSString *value = [kv[1] stringByTrimmingCharactersInSet:whitespace];
		if ([key isEqualToString:@"cuts"])   numberOfCuts = [value integerValue];
		if ([key isEqualToString:@"move"])   maxMove      = fabs([value doubleValue]);
		if ([key isEqualToString:@"rotate"]) maxRotate    = fabs([value doubleValue]);
	}

	// Process the first master of the (already interpolated) instance font,
	// honouring any include:/exclude: glyph list, as in the SDK template.
	_checkSelection = NO;
	NSString *fontMasterId = [font fontMasterAtIndex:0].id;
	if (!fontMasterId) return;
	BOOL include = NO;
	NSError *error = nil;
	NSSet *glyphNames = getIncludeExcludeGlyphListFilter(arguments, &include, font, &error);
	for (GSGlyph *glyph in font.glyphs) {
		if (glyphNames && [glyphNames containsObject:glyph.name] != include) {
			continue;
		}
		GSLayer *layer = [glyph layerForId:fontMasterId];
		if (!layer) continue;
		[self applyFilterToLayer:layer
		            numberOfCuts:numberOfCuts
		                 maxMove:maxMove
		               maxRotate:maxRotate];
	}
}

#pragma mark - Core filter logic

/**
 Apply the full CutAndShake effect to a single @p layer.
 1. Make @p numberOfCuts random angled cuts.
 2. Shift each resulting path fragment by a random vector ≤ @p maxMove.
 3. Rotate each path fragment around its own centre by ≤ @p maxRotate degrees.
 */
- (void)applyFilterToLayer:(GSLayer *)layer
              numberOfCuts:(NSInteger)numberOfCuts
                   maxMove:(CGFloat)maxMove
                 maxRotate:(CGFloat)maxRotate {

	[self randomCutLayer:layer numberOfCuts:numberOfCuts];
	[self randomMovePaths:layer    maxMove:maxMove];
	[self randomRotatePaths:layer  maxRotate:maxRotate];
}

/**
 Make @p numberOfCuts random angled cuts through @p layer.
 Each cut either spans left-to-right or top-to-bottom; both endpoints get
 independently-random positions, producing diagonal cuts (matching the Python
 version).  The cut lines extend @c kGoodMeasure beyond the layer bounds so
 they cleanly intersect all paths.
 */
- (void)randomCutLayer:(GSLayer *)layer numberOfCuts:(NSInteger)numberOfCuts {
	Class knifeTool = CutPathsToolClass();
	if (!knifeTool || layer.paths.count == 0) return;

	NSRect b = layer.bounds;
	CGFloat lowestY    = NSMinY(b) - kGoodMeasure;
	CGFloat highestY   = NSMaxY(b) + kGoodMeasure;
	CGFloat leftmostX  = NSMinX(b) - kGoodMeasure;
	CGFloat rightmostX = NSMaxX(b) + kGoodMeasure;

	for (NSInteger i = 0; i < numberOfCuts; i++) {
		NSPoint p1, p2;
		if (arc4random_uniform(2) == 0) {
			// Roughly horizontal cut: spans left-to-right, both Y values independent → angled
			p1 = NSMakePoint(leftmostX,  [self randomBetween:lowestY and:highestY]);
			p2 = NSMakePoint(rightmostX, [self randomBetween:lowestY and:highestY]);
		} else {
			// Roughly vertical cut: spans top-to-bottom, both X values independent → angled
			p1 = NSMakePoint([self randomBetween:leftmostX and:rightmostX], lowestY);
			p2 = NSMakePoint([self randomBetween:leftmostX and:rightmostX], highestY);
		}
		[knifeTool cutPathsInLayer:layer forPoint:p1 endPoint:p2];
	}
}

/**
 Translate each path in @p layer by a random vector whose magnitude
 is at most @p maxMove (distributed uniformly per axis up to maxMove/√2
 so the maximum distance equals @p maxMove).
 */
- (void)randomMovePaths:(GSLayer *)layer maxMove:(CGFloat)maxMove {
	CGFloat halfRange = maxMove / sqrt(2.0);
	for (GSPath *path in layer.paths) {
		CGFloat dx = [self randomBetween:-halfRange and:halfRange];
		CGFloat dy = [self randomBetween:-halfRange and:halfRange];
		NSAffineTransform *t = [NSAffineTransform transform];
		[t translateXBy:dx yBy:dy];
		[self applyTransform:t toNodesOfPath:path];
	}
}

/**
 Rotate each path in @p layer by a random angle in [−maxRotate, +maxRotate]
 degrees around the path's own bounding-box centre.
 */
- (void)randomRotatePaths:(GSLayer *)layer maxRotate:(CGFloat)maxRotate {
	for (GSPath *path in layer.paths) {
		NSRect  b       = path.bounds;
		CGFloat cx      = NSMidX(b);
		CGFloat cy      = NSMidY(b);
		CGFloat degrees = [self randomBetween:-maxRotate and:maxRotate];

		NSAffineTransform *t = [NSAffineTransform transform];
		[t translateXBy:cx yBy:cy];
		[t rotateByDegrees:degrees];
		[t translateXBy:-cx yBy:-cy];
		[self applyTransform:t toNodesOfPath:path];
	}
}

/**
 Apply @p transform to every node in @p path.
 This mirrors what the Python wrapper's GSPath.applyTransform() does.
 */
- (void)applyTransform:(NSAffineTransform *)transform toNodesOfPath:(GSPath *)path {
	for (GSNode *node in path.nodes) {
		node.position = [transform transformPoint:node.position];
	}
}

#pragma mark - Helpers

/** Return a uniform random CGFloat in [minimum, maximum]. */
- (CGFloat)randomBetween:(CGFloat)minimum and:(CGFloat)maximum {
	CGFloat range  = maximum - minimum;
	CGFloat random = (CGFloat)arc4random() / (CGFloat)UINT32_MAX;
	return minimum + random * range;
}

@end
